File: autoresearch-mine/scripts/chain.mjs
// Chain selection and adapter registry.
//
// Skills describe what they want done — bootstrap a project, submit a
// proposal — and never name a settlement layer. This module is the only place
// in the mine skill that knows which implementation serves which chain, so
// supporting a new chain means adding an entry here and an adapter script,
// not editing the skill or its documentation.
//
// Resolution order: explicit --chain > ARAH_CHAIN > .autoresearch/chain.json
// in the work tree > the chain recorded in a previous bootstrap > DEFAULT_CHAIN.
import fs from "node:fs";
import path from "node:path";
import process from "node:process";

export const DEFAULT_CHAIN = "solana";
export const SUPPORTED_CHAINS = ["solana", "0g"];

// operation -> chain -> { script, runner }. Adapters live beside this file.
const REGISTRY = {
  bootstrap: {
    solana: { script: "bootstrap_from_solana.mjs", runner: "node" },
    "0g": { script: "bootstrap_from_registry.py", runner: "python3" },
  },
  submitProposal: {
    solana: { script: "submit_proposal_solana.mjs", runner: "node" },
    "0g": { script: "submit_proposal.py", runner: "python3" },
  },
};

export function readJsonIfPresent(filePath) {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch {
    return null;
  }
}

export function resolveChain({ chain, workDir } = {}) {
  const candidates = [
    chain,
    process.env.ARAH_CHAIN,
    readJsonIfPresent(path.join(workDir || process.cwd(), ".autoresearch", "chain.json"))?.chain,
    readJsonIfPresent(path.join(workDir || process.cwd(), ".autoresearch", "mine", "network_state.json"))?.source,
  ];
  for (const candidate of candidates) {
    if (!candidate) continue;
    const normalized = String(candidate).toLowerCase();
    // network_state.source is "manual" when the frontier was hand-edited; that
    // says nothing about the settlement layer, so it is not a chain signal.
    if (normalized === "manual" || normalized === "registry") continue;
    if (!SUPPORTED_CHAINS.includes(normalized)) {
      throw new Error(
        `unsupported chain: ${candidate} (supported: ${SUPPORTED_CHAINS.join(", ")})`,
      );
    }
    return normalized;
  }
  return DEFAULT_CHAIN;
}

export function adapterFor(operation, chain) {
  const byChain = REGISTRY[operation];
  if (!byChain) throw new Error(`unknown operation: ${operation}`);
  const entry = byChain[chain];
  if (!entry) {
    throw new Error(`operation ${operation} is not supported on ${chain}`);
  }
  return entry;
}

export function adapterPath(scriptDir, operation, chain) {
  const entry = adapterFor(operation, chain);
  return { ...entry, path: path.join(scriptDir, entry.script) };
}

// Chain plumbing is noise during a normal run and essential when something
// fails, so it is available on request rather than removed.
export function showChainDetail() {
  return Boolean(process.env.ARAH_SHOW_CHAIN) || process.argv.includes("--show-chain");
}

export function chainDetail(message) {
  if (showChainDetail()) console.error(`[chain] ${message}`);
}

File: autoresearch-mine/scripts/chain.py
"""Chain selection, matching the resolution order used by `chain.mjs`.

Python entrypoints resolve the active settlement layer the same way the Node
entrypoints do. Keeping one order across both runtimes is the point: a miner
who sets `.autoresearch/chain.json` should not find that bootstrap honours it
and submit silently does something else.

Order, first match wins:
  explicit --chain > ARAH_CHAIN > .autoresearch/chain.json > recorded
  network_state source > DEFAULT_CHAIN
"""

from __future__ import annotations

import json
import os
from pathlib import Path

DEFAULT_CHAIN = "stellar"
SUPPORTED_CHAINS = ("stellar", "solana", "0g")

# `network_state.source` describes where the frontier came from, not where the
# project settles; "manual" and "registry" say nothing about the layer.
_NON_CHAIN_SOURCES = {"manual", "registry"}


def _read_json(path: Path) -> dict | None:
    try:
        with path.open(encoding="utf-8") as f:
            return json.load(f)
    except (OSError, json.JSONDecodeError):
        return None


def resolve_chain(explicit: str | None = None, work_dir: Path | None = None) -> str:
    root = Path(work_dir or Path.cwd())
    config = _read_json(root / ".autoresearch" / "chain.json") or {}
    state = _read_json(root / ".autoresearch" / "mine" / "network_state.json") or {}

    for candidate in (
        explicit,
        os.environ.get("ARAH_CHAIN"),
        config.get("chain"),
        state.get("source"),
    ):
        if not candidate:
            continue
        normalized = str(candidate).lower()
        if normalized in _NON_CHAIN_SOURCES:
            continue
        if normalized not in SUPPORTED_CHAINS:
            raise ValueError(
                f"unsupported chain: {candidate} (supported: {', '.join(SUPPORTED_CHAINS)})"
            )
        return normalized
    return DEFAULT_CHAIN


def show_chain_detail() -> bool:
    return bool(os.environ.get("ARAH_SHOW_CHAIN"))


def chain_detail(message: str) -> None:
    if show_chain_detail():
        import sys

        print(f"[chain] {message}", file=sys.stderr)
